bfvmm/src/hve/arch/intel_x64/ept_violation.cpp

```diff
@@ -0,0 +1,222 @@
+#include <hve/arch/intel_x64/hve.h>
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+static bool
+default_handler(
+    gsl::not_null<vmcs_t *> vmcs, ept_violation::info_t &info)
+{ bfignored(vmcs); bfignored(info); return true; }
+
+ept_violation::ept_violation(
+    gsl::not_null<eapis::intel_x64::hve *> hve
+) :
+    m_exit_handler{hve->exit_handler()}
+{
+    using namespace vmcs_n;
+
+    m_exit_handler->add_handler(
+        exit_reason::basic_exit_reason::ept_violation,
+        ::handler_delegate_t::create<ept_violation, &ept_violation::handle>(this)
+    );
+
+    this->add_read_handler(handler_delegate_t::create<default_handler>());
+    this->add_write_handler(handler_delegate_t::create<default_handler>());
+    this->add_execute_handler(handler_delegate_t::create<default_handler>());
+}
+
+ept_violation::~ept_violation()
+{
+    if (!ndebug && m_log_enabled) {
+        dump_log();
+    }
+}
+
+void
+ept_violation::add_read_handler(handler_delegate_t &&d)
+{ m_read_handlers.push_front(std::move(d)); }
+
+void
+ept_violation::add_write_handler(handler_delegate_t &&d)
+{ m_write_handlers.push_front(std::move(d)); }
+
+void
+ept_violation::add_execute_handler(handler_delegate_t &&d)
+{ m_execute_handlers.push_front(std::move(d)); }
+
+void
+ept_violation::dump_log()
+{
+    if (!m_read_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept read violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_read_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+
+    if (!m_write_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept write violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_write_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+
+    if (!m_execute_log.empty()) {
+        bfdebug_transaction(0, [&](std::string * msg) {
+            bfdebug_lnbr(0, msg);
+            bfdebug_info(0, "ept execute violation log", msg);
+            bfdebug_brk2(0, msg);
+
+            for (const auto &record : m_execute_log) {
+                bfdebug_subnhex(0, "guest virtual address", record.gva, msg);
+                bfdebug_subnhex(0, "guest physical address", record.gpa, msg);
+            }
+
+            bfdebug_lnbr(0, msg);
+        });
+    }
+}
+
+bool
+ept_violation::handle(gsl::not_null<vmcs_t *> vmcs)
+{
+    using namespace vmcs_n::exit_qualification::ept_violation;
+    auto qual = get();
```

Comment:
couple of things here:

- Its a bad idea to completely qualify this as you could end up with collisions later, and its also hard to read as I don't know if get is global or not looking at it. In this case, I would qualify vmcs_n only, and then use the actual path in the rest of your code. 
- Since you only use "qual" once, get rid of it,  and pass the value directly into the struct